[IMP] website, *: introduce WhatsApp snippet

chdh-odoo opened the pull request:
*=html_builder, website_sale

**Goal:**
Add a new WhatsApp snippet to improve social interaction between companies and website visitors. 
1. to make it easy for visitors to connect with the company as soon as they land on the website.
2. It provides a quick and familiar communication channel while also helping the business collect external user contacts.

**Preview:**
<img width="515" height="450" alt="image" src="https://github.com/user-attachments/assets/b9a93786-68da-4d67-9d06-5d78589bc3a5" /> <img width="515" height="450" alt="image" src="https://github.com/user-attachments/assets/c10e8883-f894-4ac3-a986-96473b8d8092" />


**Key Features:**
- Clicking the snippet opens a chatbot with a welcome message
- Input box lets users type queries; on send, user is redirected to personal WhatsApp with prefilled message
- Supports multiple options: size, shape, agent name, description, avatar, etc.
- Allows configuring company/WhatsApp contact number
  * Default: company's number prefilled
  * User can override with a custom number
- If the WhatsApp module (enterprise) is installed, the company can use a phone number generated from the WhatsApp Business account configuration. 
  * After that, a channel and contact card are automatically created for each new customer. (already done by WhatsApp module, not a part of this PR)
  * Note that since discuss and livechat bubbles have the same position as whatsapp snippet it's very hard to position 3 widgets at the same location so that's why we added position option so that user can change the position to left side as well.

task-4849720

shsa-odoo commented on addons/website/views/snippets/s_whatsapp.xml:
```diff
@@ -0,0 +1,49 @@
+                    <span class="notification-badge position-absolute rounded-circle"></span>
+                </div>
+                <div class="chatbox bg-white rounded-3 position-absolute shadow-lg overflow-hidden mb-2 d-none">
+                    <div class="header d-flex align-items-center p-2 text-white">
```
Add text-white to agent info only not the whole div it will make icon invisible as background itself is white.

shsa-odoo commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,101 @@
+    async apply({ editingElement }) {
+        // Only show images and icons tabs.
+        const result = await this.dependencies.media.openMediaDialog({
+            visibleTabs: ["IMAGES", "ICONS"],
```
Just keep Images for simplicity.

chdh-odoo commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,101 @@
+            whatsappNumber: this.env.getEditingElement().dataset.whatsappNumber,
+        });
+    }
+}
```
yes, i have removed that, thanks for pointing :)

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/whatsapp.edit.js:
```diff
@@ -0,0 +1,16 @@
+import { registry } from "@web/core/registry";
+import { Whatsapp } from "./whatsapp.js";
+
+export class WhatsappEdit extends Whatsapp {
```
You should use a mixin.
```js
= (I) =>
    class extends I {
```

chdh-odoo commented on addons/website/static/src/snippets/s_whatsapp/000.scss:
```diff
@@ -0,0 +1,111 @@
+.s_whatsapp {
```
follow SCSS ordering rules

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/000.scss:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    .wa-user-message {
+        field-sizing: content;
```
Not supported by Firefox (https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/field-sizing)

chdh-odoo replied:
Oh, I was leaning toward a CSS-only solution for auto-grow, but without field-sizing it could get a bit messy! (and i dont think field-sizing is going to be supported any time soon in firefox [ticket](https://bugzilla.mozilla.org/show_bug.cgi?id=1832409))

So I’ll just go with a JS approach, looks much simpler and predictable

ref: https://stackoverflow.com/questions/454202/creating-a-textarea-with-auto-resize#:~:text=OPTION%202%20(Pure%20JavaScript)

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,126 @@
+    }
+}
+
+export class WhatsappNumberAction extends BuilderAction {
+    static id = "whatsappNumber";
+    getValue({ editingElement }) {
+        return editingElement.dataset.whatsappNumber;
+    }
+    apply({ editingElement, value }) {
+        if (value) {
+            editingElement.dataset.whatsappNumber = value;
+        }
+    }
+}
```
Why do you need a custom action, you can use `DataAttributeAction` ?

chdh-odoo replied:
I initially used `DataAttributeAction`, but the requirement is to prevent users from removing the WhatsApp number once it's saved. So I need to update the value only when the new input is not empty.
```
if (value) {
        editingElement.dataset.whatsappNumber = value;
 }
```

<img width="638" height="160" alt="Image" src="https://github.com/user-attachments/assets/846798eb-b54d-4327-b868-949dbd9bd5a6" />

FrancoisGe replied:
Maybe add a small comment in the code

FrancoisGe commented on addons/html_builder/static/src/core/builder_options_plugin.js:
```diff
@@ -453,7 +453,10 @@ export class BuilderOptionsPlugin extends Plugin {
         const keepOptions =
             this.checkPredicates("should_keep_overlay_options_predicates", el) ?? false;
-        if (isInnerSnippet && isAloneInColumn && !keepOptions) {
+        const removeOptions =
+            this.checkPredicates("should_remove_overlay_options_predicates", el) ?? false;
+
+        if ((isInnerSnippet && isAloneInColumn && !keepOptions) || removeOptions) {
```
Why do you need this change ?

chdh-odoo replied:
I wanted to disable the overlay toolbar for the WhatsApp snippet, since it doesn’t make sense there. it’s not intended to be duplicated or manually moved.

<img width="219" height="101" alt="Image" src="https://github.com/user-attachments/assets/ca4f556f-9d35-48b3-853d-91e6dfb72f11" />

FrancoisGe replied:
It s not enough to use `should_keep_overlay_options_predicates` and `return false` ?

chdh-odoo replied:
Inside `hasOverlayOptions`, returning false hides the overlay options.
Right now, this is controlled by:
```
const keepOptions =
            this.checkPredicates("should_keep_overlay_options_predicates", el) ?? false;
const removeOptions =
      this.checkPredicates("should_remove_overlay_options_predicates", el) ?? false;
      
if ((isInnerSnippet && isAloneInColumn && !keepOptions) || removeOptions) {
     return false;
  }
 ```
* `removeOptions = true` → always triggers `return false`
* `keepOptions = false` → triggers `return false` **only if** `isInnerSnippet && isAloneInColumn` is also true

so that's why we introduce `should_remove_overlay_options_predicates` so even in future if anyone wants to hide overlay they can use this single predicate

FrancoisGe replied:
@sobo-odoo What do you think about it ?

chdh-odoo replied:
I already went over this part with sobo(forgot to mention here) sharing the response here so she doesn’t have to review the same thing again.

<img width="891" height="583" alt="image" src="https://github.com/user-attachments/assets/a41740ff-7ebd-4e49-8878-b8e0ef20636c" />

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option.xml:
```diff
@@ -0,0 +1,87 @@
+            title.translate="Replace Media"
+            className="'flex-grow-1'"
+            type="'success'"
+            preview="false"
```
We should move the `preview=false` in the action `replaceMedia`.

chdh-odoo replied:
what about the usage of this actions where they have added` preview="false"`? I can see two more places, one in `html_builder` and one in `website_sale`. Should we remove it from there as well?(asking in case we are targeting minimal diff)

FrancoisGe replied:
You can do it in a second commit to make it clear :)

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,126 @@
+            editingElement: editingElement,
+        });
+    }
+}
```
It s always the same code, you can create 1 custom action and use a param for the selector

FrancoisGe replied:
You can extend `toggleChatBox` in this case. It should be cleaner

chdh-odoo replied:
right good imp! thanks for suggestion :)

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/000.scss:
```diff
@@ -0,0 +1,129 @@
+        cursor: pointer;
+        caret-color: transparent;
+    }
+}
```
This css is just for the edit mode ? You should add it in an `*.edit.scss` in the `assets_inside_builder_iframe`

chdh-odoo replied:
ohh wasn't aware of that!

FrancoisGe commented on addons/website/views/snippets/s_whatsapp.xml:
```diff
@@ -0,0 +1,68 @@
+            <div class="chatbox bg-white rounded-3 position-absolute shadow-lg overflow-hidden mb-2 d-none">
+                <div class="header d-flex align-items-center p-2 text-white">
+                    <div class="o_not_editable">
+                        <img class="wa-agent-img rounded-circle" src="/website/static/src/img/snippets_demo/s_whatsapp_agent.webp" alt="agent"/>
```
You can use `o_editable_media` to enable the double click to edit the image

chdh-odoo replied:
I intentionally wrapped the image with `o_not_editable` to prevent users from changing it via double-click, since that opens sidebar options like link and size which aren’t useful here. Instead, we provided a dedicated option to change the agent’s avatar.

<img width="935" height="784" alt="Image" src="https://github.com/user-attachments/assets/16f47b8d-8a1e-49d6-8d20-d3034f607726" />

FrancoisGe commented on addons/website/static/tests/interactions/snippets/whatsapp.test.js:
```diff
@@ -0,0 +1,118 @@
+                    </button>
+                </div>
+            </div>
+            <div class="wa-messages p-2">
+                <p class="wa-agent-msg text-break w-75 bg-white rounded-3 p-2 m-0 fs-6">Hi there 👋
+                    <br/>How can I help you?
+                </p>
+            </div>
+            <div class="wa-user-input">
+                <!-- Regular input box -->
+                <div class="wa-input-box border-top align-items-end">
+                    <textarea name="wa_message" class="wa-user-message flex-grow-1 p-2 border-0" placeholder="Enter Your Message..." rows="1"></textarea>
+                    <div class="o_not_editable">
+                        <button class="wa-send border-0 p-2 text-white">
+                            <svg width="18" height="18" viewBox="0 0 24 24" fill="currentColor">
+                                <path d="M2 21l21-9L2 3v7l15 2-15 2z"/>
+                            </svg>
+                        </button>
+                    </div>
+                </div>
+                <!-- CTA mode -->
+                <div class="wa-cta-box justify-content-center p-2 border-top">
+                    <button class="wa-cta-btn d-flex align-items-center gap-2 px-5 py-3 fw-semibold rounded-4 shadow-sm border-0 text-white" >
+                        <div class="o_not_editable">
+                            <i class="fa fa-whatsapp"></i>
+                        </div>
+                        <p class="d-inline m-0">Start Conversation</p>
+                    </button>
+                </div>
+            </div>
+            <div class="wa-warning alert alert-warning d-none m-2 fs-6">
+                <i class="fa fa-info-circle me-1"></i>
+                <p class="d-inline m-0">
+                    Messaging is not available. Please contact us by another way.
+                </p>
+            </div>
+        </div>
+    </section>
+`;
+}
```
I don t know, if you know, you can directly use the real snippet. `setupWebsiteBuilderWithSnippet`

FrancoisGe replied:
My bad, you can t use it for the Interaction. We should add a way to do it. (Not in your PR but in another task)

FrancoisGe commented on addons/website/static/tests/interactions/snippets/whatsapp.test.js:
```diff
@@ -0,0 +1,118 @@
+    window.open = (url) => {
+        expect.step(`window_open ${url}`);
+    };
+}
```
You should replace `window` by `browser`

FrancoisGe replied:
And use `patchWithCleanup`

chdh-odoo replied:
I replaced window with browser in both the test and the interaction. test feels much cleaner now^^

FrancoisGe commented on addons/website/static/tests/interactions/snippets/whatsapp.test.js:
```diff
@@ -0,0 +1,118 @@
+});
+
+test("Drop Whatsapp snippet and verify warning when no number is configured", async () => {
+    await startInteractions(getWhatsappSnippet());
+    expect(".s_whatsapp").toHaveCount(1);
+    expect(".s_whatsapp .wa-warning").not.toHaveClass("d-none");
+    expect(".s_whatsapp .wa-user-input").toHaveClass("d-none");
```
Is it the interaction that adds the `d-none`?

chdh-odoo replied:
yes! during the interaction, we check whether the WhatsApp number attribute exists, and based on that, we decide whether to show a warning or display the input box.

```
const hasNumber = !!this.companyNumber;
const warningEl = this.chatbox.querySelector(".wa-warning");
const userInputEl = this.chatbox.querySelector(".wa-user-input");
// Show warning if no WhatsApp number is configured
 warningEl?.classList.toggle("d-none", hasNumber);
 userInputEl?.classList.toggle("d-none", !hasNumber);
```

FrancoisGe commented on addons/website/tests/test_snippets.py:
```diff
@@ -42,6 +42,7 @@ def test_03_snippets_all_drag_and_drop(self):
             's_snippet_group',  # Snippet groups are not snippets
             's_inline_text',
             's_drag_image_preview_test',
+            's_whatsapp',   # Fixed-position, not part of normal layout flow
```
What do you mean by `Fixed-positon` ?

chdh-odoo replied:
I initially added it because I was skeptical about the snippet making an RPC call on drop! there was a risk the tour could fail due to race conditions if the response was delayed. But now that this is handled directly in QWeb XML instead of rpc, it should be safe. we can remove our snippet from the blacklist. thanks for pointing out :)

FrancoisGe commented on addons/website/views/snippets/s_whatsapp.xml:
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+	<template id="s_whatsapp" name="WhatsApp Snippet">
+        <section class="s_whatsapp position-fixed bottom-0 o_no_save" t-att-data-whatsapp-number="env['website'].get_current_website().company_id.phone or ''">
+            <div class="position-relative d-inline-flex o_not_editable">
+                <i class="fa fa-whatsapp wa-fab rounded-circle d-flex align-items-center justify-content-center shadow-lg o_pos_right"></i>
+                <span class="notification-badge bg-danger position-absolute rounded-circle"></span>
+            </div>
+            <div class="chatbox bg-white rounded-3 position-absolute shadow-lg overflow-hidden mb-2 d-none">
+                <div class="header d-flex align-items-center p-2 text-white">
+                    <div class="o_not_editable">
```
It’s just a question: why not just add the big button and let the interaction handle the rest of the HTML?

If I understand correctly, we don’t want to be able to edit the content directly; we still want to go via the sidebar

chdh-odoo replied:
> It’s just a question: why not just add the big button and let the interaction handle the rest of the HTML?

can you elaborate more here? `renderToElement` :thinking: ? honestly here we’re just trying to prevent manual edits, and since there’s nothing dynamic going on, keeping it as static XML feels cleaner and simpler in this case

> If I understand correctly, we don’t want to be able to edit the content directly; we still want to go via the sidebar

yeah, that’s exactly why we added `content_not_editable_selectors: ".s_whatsapp"` and also wrapped media inside an `o_not_editable` parent. so users can’t directly modify that content

FrancoisGe replied:
I think you can use `renderAt` or something similar, unless you have a better solution. The advantage of this is that it minimizes the HTML tree on the page, and you can store all your attributes on the main node so that the interaction knows how to render it.

chdh-odoo replied:
Im thinking below approach which looks much cleaner and i need your pov

1. Normal mode:
Chatbox is cloned, removed from DOM, and only added when user clicks → **lazy load.**
2. Edit mode:
Chatbox is rendered immediately and stays in DOM → **easier to edit.**

This keeps the initial DOM lightweight by loading the chatbox only when the user clicks.

something like
```js
start() {
    this.chatbox = this.el.querySelector(".chatbox");
    /.../
    this.setupLazyChatbox();
}

setupLazyChatbox() {
      this.chatboxTemplateEl = this.chatbox.cloneNode(true);
      
      // Ensure chatbox is restored if interaction is destroyed/reinitialized
      this.registerCleanup(() => {
          if (!this.el.querySelector(".chatbox")) {
              this.el.append(this.chatboxTemplateEl.cloneNode(true));
          }
      });
      // Remove chatbox from DOM to enable lazy rendering
      this.chatbox.remove();
      this.chatbox = null;
  }

 toggleChat() {
      // If chatbox is not mounted, create and append it
      if (!this.chatbox && this.chatboxTemplateEl) {
          this.chatbox = this.chatboxTemplateEl.cloneNode(true);
          this.el.append(this.chatbox);
          this.chatbox.classList.remove("d-none");
          return;
      }
      // Otherwise, remove it (toggle behavior)
      this.chatbox.remove();
      this.chatbox = null;
 }
 ```
 what do u think?

chdh-odoo replied:
@FrancoisGe I've applied lazy load changes and so far it's working nicely :)

FrancoisGe replied:
I think we misunderstood each other. My idea was to keep the template as small as possible (just the green ball) and let the interaction handle everything else. As for the editable information, you can put it on the node that the interaction targets.
The advantage is that we can easily update our design without having to write update scripts.

FrancoisGe commented on addons/website/views/snippets/s_whatsapp.xml:
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+	<template id="s_whatsapp" name="WhatsApp Snippet">
+        <section class="s_whatsapp position-fixed bottom-0 o_no_save" t-att-data-whatsapp-number="env['website'].get_current_website().company_id.phone or ''">
+            <div class="position-relative d-inline-flex o_not_editable">
+                <i class="fa fa-whatsapp wa-fab rounded-circle d-flex align-items-center justify-content-center shadow-lg o_pos_right"></i>
+                <span class="notification-badge bg-danger position-absolute rounded-circle"></span>
+            </div>
```
A designer will revisit the block's design to bring it more in line with this one
<img width="317" height="316" alt="Image" src="https://github.com/user-attachments/assets/d99da0d1-7bf4-47d8-a7c8-af11f8a6bae4" />

chdh-odoo replied:
hmm looks awesome!

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/whatsapp.js:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    toggleChat() {
+        this.chatbox.classList.toggle("d-none");
```
Should this change be recorded in the history or not ?

chdh-odoo replied:
Toggling the chatbox (open/close) is a UI-only interaction, not a structural or content change to the page. i think history steps should be reserved for actions that affect the saved state of the website. because after save we anyway going to close the chatbox.

FrancoisGe replied:
In your case here, the change isn't being ignored, so it will carry over to the next step. So when you undo, it will toggle it again.

chdh-odoo replied:
currently i think it's being ignored, im not able to toggle it again when i do undo

Ref: https://www.loom.com/share/0253b6315e7841f5adb5bfd0316f5684

livaios commented on addons/website_sale/static/src/website_builder/product_image_option.xml:
```diff
@@ -8,7 +8,6 @@
             title.translate="Replace image"
             className="'flex-grow-1'"
             type="'success'"
-            preview="false"
```
Why was this change needed in this task?

chdh-odoo replied:
we’re using the same `action="replaceMedia"` for the whatsapp snippet, and noticed we have to set `preview="false"` every time in XML. instead of repeating it everywhere, it’s cleaner to just set it to false in the setup itself.

livaios replied:
Thanks for the explanation 🙂

FrancoisGe commented on addons/website/views/snippets/s_whatsapp.xml:
```diff
@@ -2,63 +2,26 @@
+        <section
+            class="s_whatsapp position-fixed bottom-0 o_no_save"
+            t-att-data-whatsapp-number="env['website'].get_current_website().company_id.phone or ''"
+            data-agent-name="Jane Doe"
+            data-agent-description="Online"
+            data-agent-message="Hi there 👋 how can I help you?"
+            data-agent-avatar-src="/website/static/src/img/snippets_demo/s_whatsapp_agent.webp"
```
You can move each `data-` to the button

chdh-odoo replied:
Since translation depends on input fields with `o_translatable_input_hidden` class, it makes more sense to rely on it entirely and remove those extra attributes.

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
You don t need to change the content in the dom. It should be automatic when you edit the dataset on the node. (Interaction should be rerender)

chdh-odoo replied:
I initially tried that approach, but ran into an issue where the avatar option disappears. The problem is that when the interaction refreshes after editing fields like name, description, or message, the `.wa-agent-img` element gets temporarily re-rendered. During that moment, the builder can’t find the target, so the Avatar option vanishes.

This applyTo behavior seems to be a generic issue. I tested it on other snippets as well. To check, Inside countdown snippet, I added a “Test Image” option using the standard replaceMedia action with applyTo="'.s_countdown_test_img'". When an interaction refresh is triggered through options like Due Date or Display, the Test Image option disappears. Video [Link](https://www.loom.com/share/d3f4b865ac134b69b5ab6632602347f4)

As a workaround I think we can fix that by not of relying on `applyTo`, we can give target inside a custom builder action. That way, the option won’t depend on a specific DOM node being present, and it will remain stable even if the chatbox is re-rendered multiple times due to interaction refreshes.

FrancoisGe replied:
@chdh-odoo  It seems wrong to me that we have to edit a DOM element that’s been manipulated by an interaction. That means the DOM element will be destroyed by the interaction.
Can you push a commit where I’ll see the issue and let me know how to fix it?

FrancoisGe replied:
In my opinion, you just need to edit the node with all the `data-` attributes, and the interaction will then render with the correct information.

chdh-odoo replied:
> It seems wrong to me that we have to edit a DOM element that’s been manipulated by an interaction.

Yes, this can be handled by updating a data attribute. If we add `dataAttributeAction` to the chatbox options, it will automatically trigger an interaction refresh.

but.. since data attributes cannot be translated, we introduced input fields to act as the source of truth. That means we need to keep those input values in sync. Given that, also writing the same values to data attributes feels redundant, which is why I avoided using `dataAttributeAction`.

```xml
  <div class="s_whatsapp_translation_inputs">
        <input type="hidden" class="o_translatable_input_hidden d-block mb-1 w-100" data-name="Name" name="agent_name" value="Jane Doe"/>
        <input type="hidden" class="o_translatable_input_hidden d-block mb-1 w-100" data-name="Description" name="agent_description" value="Online"/>
        <input type="hidden" class="o_translatable_input_hidden d-block mb-1 w-100" data-name="Default Message" name="agent_message" value="Hi there 👋 how can I help you?"/>
        <div class="wa-input-box">
            <input type="hidden" class="o_translatable_input_hidden d-block mb-1 w-100" data-name="Placeholder" name="message_placeholder" value="Enter Your Message..."/>
        </div>
        <div class="wa-cta-box">
            <input type="hidden" class="o_translatable_input_hidden d-block mb-1 w-100" data-name="CTA Label" name="cta_label" value="Start Conversation"/>
        </div>
    </div>
 ```
            
> Can you push a commit where I’ll see the issue and let me know how to fix it?

Sure I'll add the same option in countdown snippet and you can see the issue by following steps in this video [link](https://www.loom.com/share/d3f4b865ac134b69b5ab6632602347f4)

chdh-odoo replied:
Pushed the fourth commit where you can reproduce the issue for countdown snippet 

In general when the `applyTo` target is an img node, the sidebar loses its context during interaction refresh and becomes invisible. as a workaround, I removed `applyTo` and instead passed the target through a custom builder action.

FrancoisGe replied:
I wanted you to enable the settings in your WhatsApp snippet. I'm not sure I understand what you mean by “inputs in translation mode.”

chdh-odoo replied:
> I wanted you to enable the settings in your WhatsApp snippet.

oh then that is easy to reproduce just use `applyTo` with  standard `replaceMedia` action

 means instead of `action="'replaceAgentAvatar'"` add
```js
action="'replaceMedia'" applyTo="'.wa-agent-img'"
```

> I'm not sure I understand what you mean by “inputs in translation mode.”

Custom attributes are not translatable by default, so we need a node that can store values per language. That’s why we introduced .s_whatsapp_translation_inputs, with inputs for each chatbox text such as name, description, placeholder, etc. donation snippet also follows the very similar approach.

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/whatsapp.edit.js:
```diff
@@ -3,8 +3,23 @@ import { Whatsapp } from "./whatsapp.js";
+            if (shouldShow) {
+                delete this.el.dataset.shouldShowChatbox;
+            }
+        }
```
Why you don't want to restart it ?

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/whatsapp.js:
```diff
@@ -2,6 +2,11 @@ import { registry } from "@web/core/registry";
+const DEFAULT_AGENT_NAME = "Jane Doe";
+const DEFAULT_AGENT_DESCRIPTION = "Online";
+const DEFAULT_AGENT_MESSAGE = "Hi there 👋 how can I help you?";
+const DEFAULT_AGENT_AVATAR_SRC = "/website/static/src/img/snippets_demo/s_whatsapp_agent.webp";
```
Why did you move it in const ?  It s already in the dom of the snippet.
You should use `_t()`

chdh-odoo replied:
We can remove those variables since we now depend on input values, as they will always be present once set by the user.

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/whatsapp.js:
```diff
@@ -17,38 +22,50 @@ export class Whatsapp extends Interaction {
+    }
+
+    mountChatbox({ show = false } = {}) {
+        if (!this.chatbox) {
```
You double check ? (toggle)

chdh-odoo commented:
Hello @FrancoisGe 

Here's the change log since last review :)

1. A hidden input with `o_translatable_input_hidden` is added, containing chatbox texts such as name, description, placeholder, etc. These inputs are used for translation and act as the source of truth. When we change options from sidebar, we update the values of these inputs as well. The same applies in translation mode, and then interactions automatically fetch values from these inputs to render the content.

2. Removed data attributes since we dont need them(we're already keeping `o_translatable_input_hidden` inputs in sync).

3. The chatbox is hidden in translation mode since we already show a list of inputs where users can edit the translation of the chatbox content, following the same approach as the donation snippet.

4. Not sure if we can add test coverage for chatbox content options, since HOOT tests for the Website Builder do not activate interactions as of now when using `setupWebsiteBuilder`. I have added coverage for other options that do not rely on interactions.

Note that I will squash the third commit into the first one at the end and remove the fourth one(as it was only meant to demonstrate that the issue with `applyTo` is generic.)

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option.xml:
```diff
@@ -0,0 +1,85 @@
+<templates xml:space="preserve">
+
+	<t t-name="website.WhatsappOption">
+		<t t-if="!state.whatsappNumber">
+			<div class="alert alert-info">
+                A Contact Number is essential, as WhatsApp chat redirection depends on a valid number.
+            </div>
+		</t>
```
I think it s not possible to have this situation ? When remove the number, I always keep the previous value or a default value.

chdh-odoo replied:
It's possible when company (res.company) also dont have phone number set. in that case the contact number option will be empty so somehow we have to show that "you must enter number here"

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,128 @@
+            WhatsappNumberAction,
+        },
+        system_attributes: ["data-should-show-chatbox"],
+        content_not_editable_selectors: ".s_whatsapp",
```
Do you still need this line ? (Now it s an Interaction)

chdh-odoo replied:
hmm right! we can remove this since it's not needed now. thanks for pointing out :)

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,128 @@
+        const rootWhatsappEl = editingElement.closest(".s_whatsapp");
+        if (rootWhatsappEl) {
+            rootWhatsappEl.dataset.shouldShowChatbox = "true";
+            rootWhatsappEl.querySelector(".chatbox")?.classList.remove("d-none");
```
Why do we have two things that seem to do the same thing? The `d-none` class, which determines whether or not the chatbot is displayed, and `shouldShowChatbox`, which seems to do the same thing. Can't we just merge the two?

chdh-odoo replied:
The second condition is for the “Show CTA” option. Right now, changing it doesn’t trigger the interaction so that's why I had to manully remove this d-none but...

I think Instead of using classAction for show CTA option, we should switch to dataAttributeAction so it automatically triggers the interaction refresh and we can remove this manual DOM manipulation.

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,128 @@
+    }
+}
+
+export class replaceAgentAvatarAction extends BuilderAction {
```
```suggestion
export class ReplaceAgentAvatarAction extends BuilderAction {
```

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,128 @@
+                whatsappEl.dataset.shouldShowChatbox = "true";
+                whatsappEl.querySelector(".chatbox")?.classList.remove("d-none");
+            }
+        }
```
We can t do it in `ReplaceAgentAvatarAction` ?

chdh-odoo replied:
`ReplaceAgentAvatarAction` only runs when the user clicks “replace avatar.” After that, once the media dialog opens, we don’t have a way to know which image the user selected, and we need the source of the newly chosen media in order to set as data attribute.

FrancoisGe replied:
It looks like a limitation of the editor. I guess we'll just have to live with it.

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,128 @@
+            // the chatbox input for user interaction.
+            const warningEl = editingElement.querySelector(".wa-warning");
+            const userInputEl = editingElement.querySelector(".wa-user-input");
+            warningEl?.classList.add("d-none");
```
We only use this warning in edit mode ? Or it s visible on the website ?

lebl-odoo replied:
Hello @chdh-odoo , regarding the warning, could you check we only support phone number. You don't need to create a huge regex, but at least we should prevent some char (like a-zA-Z, but double check)

chdh-odoo replied:
yes it's visible in website as well since without phone number the redirection will not work.

warning message contains "Messaging is not available. Please contact us by another way." instead of showing chatbox

chdh-odoo replied:
@lebl-odoo I’ve already added validation during **redirection**, so any characters entered mistakenly will be ignored. Do you still want me to show an error notification?

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/whatsapp_chatbox.xml:
```diff
@@ -0,0 +1,52 @@
+                <div class="o_not_editable">
+                    <button class="btn wa-close-btn p-2 text-white" aria-label="Close chat">
+                        <i class="fa fa-times"></i>
+                    </button>
+                </div>
+            </div>
+            <div class="wa-messages p-2">
+                <p class="wa-agent-msg text-break w-75 bg-white rounded-3 p-2 m-0 fs-6" t-esc="agentMessage"/>
+            </div>
+            <div class="wa-user-input">
+                <!-- Regular input box -->
+                <div class="wa-input-box border-top align-items-end">
+                    <textarea name="wa_message" class="wa-user-message flex-grow-1 p-2 border-0" t-att-placeholder="messagePlaceholder" rows="1"></textarea>
+                    <div class="o_not_editable">
+                        <button class="wa-send border-0 p-2 text-white" aria-label="Send message">
+                            <svg width="18" height="18" viewBox="0 0 24 24" fill="currentColor">
+                                <path d="M2 21l21-9L2 3v7l15 2-15 2z"/>
+                            </svg>
+                        </button>
+                    </div>
+                </div>
+                <!-- CTA mode -->
+                <div class="wa-cta-box justify-content-center p-2 border-top">
+                    <button class="wa-cta-btn d-flex align-items-center gap-2 px-5 py-3 fw-semibold rounded-4 shadow-sm border-0 text-white" t-att-aria-label="ctaLabel">
+                        <div class="o_not_editable">
+                            <i class="fa fa-whatsapp"></i>
+                        </div>
+                        <p class="m-0" t-esc="ctaLabel"/>
+                    </button>
+                </div>
+            </div>
+            <div class="wa-warning alert alert-warning d-none m-2 fs-6">
+                <i class="fa fa-info-circle me-1"></i>
+                <p class="d-inline m-0">
+                    Messaging is not available. Please contact us by another way.
+                </p>
+            </div>
+        </div>
+    </t>
+</templates>
```
Are you sure we want to version this html ? Because it s not in the dom of the website anymore. (It will be only use by the Interaction ?)

chdh-odoo replied:
That’s what I assumed, but I noticed an example with `.s_countdown.end_redirect_message`, `s_searchbar.autocomplete.skeleton.loader`  where versioning is still applied even though it’s rendered via an interaction, so it made me a bit unsure.

chdh-odoo replied:
I checked more and I think versioning on the frontend template isn’t necessary. I can just use `whatsapp_chatbox.xml`

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/whatsapp.edit.js:
```diff
@@ -0,0 +1,52 @@
+            super.start();
+            const shouldShow = this.el.dataset.shouldShowChatbox === "true";
+            this.mountChatbox({ show: shouldShow });
+            this.registerCleanup(() => {
+                // We only remove the dataset when the interaction is fully
+                // destroyed, not during refresh, so the open state can persist
+                // across re-renders.
+                if (!this.services["public.interactions"].isRefreshing) {
+                    delete this.el.dataset.shouldShowChatbox;
+                }
```
Since it's a technical attribute, it will be cleared when you save. Isn't that enough?

chdh-odoo replied:
Even though shouldShowChatbox is a technical attribute, it is not cleared when the snippet is saved, so that's why I removed during cleanup.

FrancoisGe replied:
My bad, it s still in progress https://www.odoo.com/odoo/project/1695/tasks/5156767

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/whatsapp.edit.js:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        toggleChat() {
+            this.chatbox.classList.toggle("d-none");
```
We should probably toggle `shouldShowChatbox` ?

FrancoisGe commented on addons/website/static/src/snippets/s_whatsapp/whatsapp.js:
```diff
@@ -0,0 +1,103 @@
+        }
+    }
+
+    autoGrow(ev) {
+        const textAreaEl = ev.target;
+        textAreaEl.style.height = "auto";
+        textAreaEl.style.height = `${textAreaEl.scrollHeight}px`;
+    }
```
<img width="343" height="292" alt="Image" src="https://github.com/user-attachments/assets/4d3895e0-b1a4-4ed2-87df-bcb69156801e" />
Style issue for the send button

chdh-odoo replied:
I can’t stretch the send button horizontally since it would look distorted, so I followed the standard WhatsApp behavior instead.

Here's how it look on whatsapp: https://www.loom.com/share/dd44efd8e4e24ddbabce5d51dc106a93

But let me know if we want something diff

FrancoisGe replied:
I probably get that weird feeling because, in your case, the button follows the shape of the popup, and it's not just a simple circle.

chdh-odoo replied:
Based on the [design](https://github.com/odoo/odoo/pull/223142#discussion_r3064708417) you shared earlier, the designer is going to make it rounded anyway, so I think we're good here.

FrancoisGe replied:
Could you try to the same design ? (you can do it in another commit)

chdh-odoo replied:
I have made it rounded and looks much better now^^ I've kept in same commit since it's just 2 lines of css

```
border-radius: 50%;
margin: 6px;
```

video: https://www.loom.com/share/bf0909158091468994044ec91cfc507d

FrancoisGe commented:
@chdh-odoo 

 one detail, when you are in edit mode, the input ("Enter your message") should be selectable. It s a small ux imp)

FrancoisGe commented:
@chdh-odoo You can clear your commits :)

lebl-odoo commented:
@FrancoisGe , @chdh-odoo ,

The warning makes sense if phone number is not set or wrong.

We could create a follow up task for multiple phone numbers, but we could wait for some feedbacks.

chdh-odoo commented:
> one detail, when you are in edit mode, the input ("Enter your message") should be selectable. It s a small ux imp)

I intentionally disabled it based on previous reviews, since users can’t type or send messages in edit mode anyway, but I’ve made it selectable :relaxed: 

apart from this, I’ve incorporated other suggestions as well.

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -0,0 +1,126 @@
+export class ReplaceAgentAvatarAction extends BuilderAction {
+    static id = "replaceAgentAvatar";
+    static dependencies = ["media_website"];
+
+    setup() {
+        this.preview = false;
+        this.canTimeout = false;
+    }
+
+    async apply({ editingElement }) {
+        // The avatar is created dynamically and gets re-rendered by interaction
+        // so the element may not exist when the builder tries to target it.
+        // That’s why we didn’t use the standard replaceMedia action.
+        const avatarEl = editingElement.querySelector(".wa-agent-img");
+        if (avatarEl) {
+            await this.dependencies.media_website.replaceMedia(avatarEl);
+        }
+    }
```
We shouldn't target the Interaction content. You should target the data attribut. And when we change the data attribute the interaction will be reload.

chdh-odoo replied:
Hmm In that case we can't use the standard replaceMedia action here because it requires passing the editing node, but we still have alternative.

We can call `openMediaDialog` directly to retrieve the newly selected image src and control which tab opens. This also allows us to remove `replace_media_dialog_params_processors` and `on_media_replaced_handlers`, so the benefits seem worthwhile.

FrancoisGe replied:
Yes you can try this solution in another commit. I want to check the diff :)

chdh-odoo replied:
Sure Here's the [commit](https://github.com/odoo/odoo/pull/223142/changes/93f1a064e59f93d49f9725cf2823a7374280bb1e?new_files_changed=true)

FrancoisGe replied:
Nice, I prefer this solution. It s more centralize and easier to read :)

chdh-odoo replied:
cool, I've cleaned the commits

FrancoisGe commented:
> > one detail, when you are in edit mode, the input ("Enter your message") should be selectable. It s a small ux imp)
> 
> I intentionally disabled it based on previous reviews, since users can’t type or send messages in edit mode anyway, but I’ve made it selectable ☺️
> 
> apart from this, I’ve incorporated other suggestions as well.

My bad, I meant to say that you shouldn't be able to select it or focus on it. I think it was focusable.

chdh-odoo commented:
@FrancoisGe 

> My bad, I meant to say that you shouldn't be able to select it or focus on it. I think it was focusable.

I’ve disabled pointer events so the user can’t interact with it in edit mode.

chdh-odoo commented:
@FrancoisGe ping :relaxed:

FrancoisGe commented:
> @FrancoisGe ping ☺️

I just post 2 small message. After, we need to await the designer PR.

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option.xml:
```diff
@@ -28,6 +28,7 @@
 			<BuilderButton
 				action="'replaceAgentAvatar'"
 				title.translate="Replace Media"
+				preview="false"
```
You can keep in the js.

chdh-odoo replied:
This action is not going to be re-used anywhere else except this snippet should't we keep this be in xml then? 

(I still confuse in which case we should prefer js over xml)

FrancoisGe replied:
For me, by default, if an action can't be previewed, then it goes in the JS. If it can be previewed, then it goes in the XML. This is more from a technical standpoint. For example, “replaceAgentAvatar” will always open a dialog => it can never be previewed

dhruv2x replied:
hmm makes sense! thank you for explanation :)

FrancoisGe commented on addons/website/static/src/builder/plugins/options/whatsapp_option_plugin.js:
```diff
@@ -80,23 +56,24 @@ export class ShowChatBoxAction extends BuilderAction {
-    setup() {
-        this.preview = false;
-        this.canTimeout = false;
-    }
```
You should keep this part and add `super.setup()`

FrancoisGe commented:
@robodoo r+ rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.
